File: packages/subscription-pricing/src/annual-prepay.ts
import type { SubscriptionPlan } from '@quokkaq/shared-types';

/** True when the plan lists a 12-month prepay option (monthly interval + discount or fixed effective monthly). */
export function planSupportsAnnualPrepay(plan: SubscriptionPlan): boolean {
  if (plan.isFree || plan.interval !== 'month' || plan.price <= 0) {
    return false;
  }
  const d = plan.annualPrepayDiscountPercent;
  const m = plan.annualPrepayPricePerMonth;
  if (d != null && m != null) {
    return false;
  }
  if (d != null) {
    return Number.isInteger(d) && d >= 1 && d <= 100;
  }
  if (m != null) {
    return Number.isFinite(m) && m > 0;
  }
  return false;
}

/** Total minor units charged once per year in Stripe for annual prepay checkout. */
export function annualPrepayStripeYearlyUnitAmountMinor(
  plan: SubscriptionPlan
): number | null {
  if (!planSupportsAnnualPrepay(plan)) {
    return null;
  }
  const ppm = plan.annualPrepayPricePerMonth;
  if (ppm != null) {
    return Math.trunc(ppm * 12);
  }
  const pct = plan.annualPrepayDiscountPercent;
  if (pct == null) {
    return null;
  }
  return Math.trunc((plan.price * 12 * (100 - pct)) / 100);
}

/** Effective monthly price in minor units when paying for 12 months (display). */
export function annualPrepayEffectiveMonthlyMinor(
  plan: SubscriptionPlan
): number | null {
  const yearly = annualPrepayStripeYearlyUnitAmountMinor(plan);
  if (yearly == null) {
    return null;
  }
  return Math.trunc(yearly / 12);
}

/** Approximate savings vs 12× list monthly (minor units); null if not applicable. */
export function annualPrepaySavingsMinorUnits(
  plan: SubscriptionPlan
): number | null {
  const yearly = annualPrepayStripeYearlyUnitAmountMinor(plan);
  if (yearly == null) {
    return null;
  }
  return Math.max(0, plan.price * 12 - yearly);
}

/**
 * Discount percent for marketing/UI: explicit annualPrepayDiscountPercent, or an
 * equivalent rounded % when only annualPrepayPricePerMonth is set (vs list monthly).
 */
export function annualPrepayDisplayDiscountPercent(
  plan: SubscriptionPlan
): number | null {
  if (!planSupportsAnnualPrepay(plan)) {
    return null;
  }
  const d = plan.annualPrepayDiscountPercent;
  if (d != null) {
    return d;
  }
  const ppm = plan.annualPrepayPricePerMonth;
  if (ppm != null && plan.price > 0) {
    const raw = (1 - ppm / plan.price) * 100;
    return Math.max(0, Math.min(99, Math.round(raw)));
  }
  return null;
}

File: apps/backend/internal/subscriptionplan/annual_prepay.go
package subscriptionplan

import (
	"errors"
	"fmt"

	"quokkaq-go-backend/internal/models"
)

var (
	ErrAnnualPrepayBothFields    = errors.New("annual prepay: set only one of annualPrepayDiscountPercent or annualPrepayPricePerMonth")
	ErrAnnualPrepayInterval      = errors.New("annual prepay options require interval month")
	ErrAnnualPrepayFreeOrZero    = errors.New("annual prepay options require a non-free plan with positive price")
	ErrAnnualPrepayDiscount      = errors.New("annualPrepayDiscountPercent must be between 1 and 100")
	ErrAnnualPrepayPricePerMo    = errors.New("annualPrepayPricePerMonth must be positive")
	ErrAnnualPrepayNotConfigured = errors.New("plan does not support annual prepay checkout")
)

// HasAnnualPrepayConfig reports whether the plan advertises a 12-month prepay option (monthly catalog only).
func HasAnnualPrepayConfig(p *models.SubscriptionPlan) bool {
	if p == nil {
		return false
	}
	if p.IsFree || p.Interval != "month" || p.Price <= 0 {
		return false
	}
	d := p.AnnualPrepayDiscountPercent
	m := p.AnnualPrepayPricePerMonth
	if d != nil && m != nil {
		return false
	}
	if d != nil {
		return *d >= 1 && *d <= 100
	}
	if m != nil {
		return *m > 0
	}
	return false
}

// AnnualPrepayYearlyUnitAmountMinor returns the Stripe subscription unit amount for one yearly billing period (minor units).
func AnnualPrepayYearlyUnitAmountMinor(p *models.SubscriptionPlan) (int64, error) {
	if !HasAnnualPrepayConfig(p) {
		return 0, ErrAnnualPrepayNotConfigured
	}
	if p.AnnualPrepayPricePerMonth != nil {
		v := *p.AnnualPrepayPricePerMonth
		if v <= 0 {
			return 0, ErrAnnualPrepayPricePerMo
		}
		return v * 12, nil
	}
	pct := *p.AnnualPrepayDiscountPercent
	if pct < 1 || pct > 100 {
		return 0, ErrAnnualPrepayDiscount
	}
	// Integer math: 12 months at list price, then apply discount.
	return p.Price * 12 * int64(100-pct) / 100, nil
}

// ValidateAnnualPrepayFields checks mutual exclusion and consistency with interval / price / isFree.
func ValidateAnnualPrepayFields(interval string, isFree bool, price int64, disc *int, ppm *int64) error {
	hasDisc := disc != nil
	hasPPM := ppm != nil
	if hasDisc && hasPPM {
		return ErrAnnualPrepayBothFields
	}
	if !hasDisc && !hasPPM {
		return nil
	}
	if interval != "month" {
		return ErrAnnualPrepayInterval
	}
	if isFree || price <= 0 {
		return ErrAnnualPrepayFreeOrZero
	}
	if hasDisc {
		if *disc < 1 || *disc > 100 {
			return ErrAnnualPrepayDiscount
		}
	}
	if hasPPM && *ppm <= 0 {
		return ErrAnnualPrepayPricePerMo
	}
	return nil
}

// CheckoutLineForBilling returns Stripe line fields for checkout, or nil to use the plan as-is.
func CheckoutLineForBilling(p *models.SubscriptionPlan, billingPeriod string) (*CheckoutSubscriptionLine, error) {
	bp := billingPeriod
	if bp == "" || bp == "month" {
		return nil, nil
	}
	if bp != "annual" {
		return nil, fmt.Errorf("unknown billingPeriod: %s", billingPeriod)
	}
	amt, err := AnnualPrepayYearlyUnitAmountMinor(p)
	if err != nil {
		return nil, err
	}
	name := p.Name
	if name == "" {
		name = p.Code
	}
	return &CheckoutSubscriptionLine{
		UnitAmount:  amt,
		Interval:    "year",
		ProductName: name + " (annual)",
	}, nil
}

// CheckoutSubscriptionLine overrides plan price/interval for Stripe checkout.
type CheckoutSubscriptionLine struct {
	UnitAmount  int64
	Interval    string
	ProductName string
}
